numpy/core/src/multiarray/mapping.c

```diff
@@ -235,29 +210,85 @@ prepare_index(PyArrayObject *self, PyObject *index,
 
         if (make_tuple) {
             /* We want to interpret it as a tuple, so make it one */
-            index = PySequence_Tuple(index);
-            if (index == NULL) {
-                return -1;
+            index_as_tuple = PySequence_Tuple(index);
+            if(index_as_tuple == NULL) {
+                return NULL;
             }
         }
     }
 
-    /* If the index is not a tuple, handle it the same as (index,) */
-    if (!PyTuple_CheckExact(index)) {
-        obj = index;
-        index_ndim = 1;
+    /* If the index is not a tuple, conver it into (index,) */
+    if (!make_tuple && !PyTuple_CheckExact(index)) {
+        make_tuple = 1;
+        index_as_tuple = Py_BuildValue("(O)", index);
     }
+    /* Otherwise, check if the tuple is too long */
     else {
-        n = PyTuple_GET_SIZE(index);
+        n = PyTuple_GET_SIZE(index_as_tuple);
         if (n > NPY_MAXDIMS * 2) {
```

Comment:
Well if you pass in a finished tuple you have to get the same error at some point.